File: app/src/main/java/com/nicokla/musicos2/navigation/UserAdapter.java
package com.nicokla.musicos2.navigation;

import android.content.res.Resources;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.Query;

import com.nicokla.musicos2.Firebase.UserFirestore;
import com.nicokla.musicos2.MySongsFrag.FirestoreAdapter;
import com.nicokla.musicos2.databinding.SongItemWithAuthorBinding;
import com.nicokla.musicos2.databinding.UserItemBinding;

class UserAdapter extends FirestoreAdapter<UserAdapter.ViewHolder> {

  public interface OnUserSelectedListener {

    void onUserSelected(DocumentSnapshot song);

  }

  private UserAdapter.OnUserSelectedListener mListener;

  public UserAdapter(Query query, UserAdapter.OnUserSelectedListener listener) {
    super(query);
    mListener = listener;
  }

  @Override
  public UserAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    return new UserAdapter.ViewHolder(UserItemBinding.inflate(
            LayoutInflater.from(parent.getContext()), parent, false));
  }

  @Override
  public void onBindViewHolder(UserAdapter.ViewHolder holder, int position) {
    holder.bind(getSnapshot(position), mListener);
  }

  static class ViewHolder extends RecyclerView.ViewHolder {

    private UserItemBinding binding;

    public ViewHolder(@NonNull UserItemBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }

    public ViewHolder(View itemView) {
      super(itemView);
    }

    public void bind(final DocumentSnapshot snapshot,
                     final UserAdapter.OnUserSelectedListener listener) {

      UserFirestore user = snapshot.toObject(UserFirestore.class);
      Resources resources = itemView.getResources();

      binding.userName.setText(user.name);

      // Click listener
      binding.userView.setOnClickListener(new View.OnClickListener() {
        @Override
        public void onClick(View view) {
          if (listener != null) {
            listener.onUserSelected(snapshot);
          }
        }
      });
    }

  }
}

File: app/src/main/java/com/nicokla/musicos2/algolia/user/UserAdapter.kt
package com.nicokla.musicos2.algolia.user

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
//import com.algolia.instantsearch.guides.R
//import com.algolia.instantsearch.guides.gettingstarted.User
//import com.algolia.instantsearch.guides.gettingstarted.UserViewHolder
import com.nicokla.musicos2.R
import com.nicokla.musicos2.algolia.song.SongViewHolder
import com.nicokla.musicos2.databinding.SongItemWithAuthorBinding
import com.nicokla.musicos2.databinding.UserItemBinding


class UserAdapter(private val cellClickListener: CellClickListener)
    : PagedListAdapter<User, UserViewHolder>(UserAdapter) {


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
    UserViewHolder(
            // Inflate generated item binding class
            UserItemBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false
            )
    )

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        val user = getItem(position)

        if (user != null) {
            holder.bind(user)
            holder.itemView.setOnClickListener {
                cellClickListener.onCellClickListener(user)
            }
        }
    }

    companion object : DiffUtil.ItemCallback<User>() {
        override fun areItemsTheSame(
                oldItem: User,
                newItem: User
        ): Boolean {
            return oldItem::class == newItem::class
        }

        override fun areContentsTheSame(
                oldItem: User,
                newItem: User
        ): Boolean {
            return oldItem.name == newItem.name
        }
    }
}